pkg/master/controller.go

```diff
@@ -226,19 +302,19 @@ func (c *Controller) CreateOrUpdateMasterServiceIfNeeded(serviceName string, ser
 }
 
 // ReconcileEndpoints sets the endpoints for the given apiserver service (ro or rw).
-// ReconcileEndpoints expects that the endpoints objects it manages will all be
-// managed only by ReconcileEndpoints; therefore, to understand this, you need only
-// understand the requirements and the body of this function.
-//
-// Requirements:
-//  * All apiservers MUST use the same ports for their {rw, ro} services.
-//  * All apiservers MUST use ReconcileEndpoints and only ReconcileEndpoints to manage the
-//      endpoints for their {rw, ro} services.
-//  * All apiservers MUST know and agree on the number of apiservers expected
-//      to be running (c.masterCount).
-//  * ReconcileEndpoints is called periodically from all apiservers.
-//
+// It does this by listing keys in an etcd directory.  Each key is expected to have a
+// TTL of R+n, where R is the refresh interval at which this function is called, and
+// n is some small value.  If an apiserver goes down, it will fail to refresh its key's
+// TTL and the key will expire.  ReconcileEndpoints will notice that the endpoints object
+// is different from the directory listing, and update the endpoints object accordingly.
 func (c *Controller) ReconcileEndpoints(serviceName string, ip net.IP, endpointPorts []api.EndpointPort, reconcilePorts bool) error {
+
+	// First, refresh the TTL on our key
```

Comment:
I would highlight the point that "First" means that we update our ttl in every iteration, independently whether any error or update conflict happens below. This makes sure that at least some of the master will add our endpoint.